src/fireedge/src/server/routes/api/auth/utils.js

```diff
@@ -0,0 +1,669 @@
+ */
+const connectOpennebula = (usr = '', pss = '') => {
+  const connectUser = usr || user
+  const connectPass = pss || pass
+
+  return nodeConnect(connectUser, connectPass)
+}
+
+/**
+ * Updater http request.
+ *
+ * @param {string} code - http code
+ */
+const updaterResponse = (code) => {
+  if ('id' in code && 'message' in code && res?.locals?.httpCode) {
+    res.locals.httpCode = code
+  }
+}
+
+/**
+ * Validate 2FA.
+ *
+ * @param {object} informationUser - user data
+ * @returns {boolean} return if data is valid
+ */
+const validate2faAuthentication = (informationUser) => {
+  let rtn = false
+  const template = informationUser?.TEMPLATE
+  if (
+    template?.SUNSTONE?.[default2FAOpennebulaVar] ||
+    template?.FIREEDGE?.[default2FAOpennebulaVar]
+  ) {
+    /*********************************************************
+     * Validate 2FA
+     *
+     * Gives priority to the 2FA created directly in the fireedge and in case it is not configured, uses the 2FA created from the sunstone.
+     *********************************************************/
+
+    if (tfatoken.length <= 0) {
+      updaterResponse(httpResponse(accepted, { id: informationUser?.ID }))
+    } else {
+      const secret =
+        template.FIREEDGE?.[default2FAOpennebulaVar] ||
+        template.SUNSTONE?.[default2FAOpennebulaVar]
+      if (!check2Fa(secret, tfatoken)) {
+        updaterResponse(httpResponse(unauthorized, '', 'invalid 2fa token'))
+      } else {
+        rtn = true
+      }
+    }
+  } else {
+    /*********************************************************
+     * Without 2FA
+     *********************************************************/
+
+    rtn = true
+  }
+
+  return rtn
+}
+
+/**
+ * Generate a JWT.
+ *
+ * @param {string} token - opennebula token
+ * @param {object} informationUser - user data
+ */
+const genJWT = (token, informationUser) => {
+  if (
+    token?.time &&
+    token?.token &&
+    informationUser?.ID &&
+    informationUser?.NAME
+  ) {
+    const { ID: id, TEMPLATE: userTemplate, NAME: username } = informationUser
+    const jwt = createJWT({ id, user: username, token: token.token })
+
+    if (jwt) {
+      const rtn = { token: jwt, id }
+      if (userTemplate?.SUNSTONE?.LANG) {
+        rtn.language = userTemplate.SUNSTONE.LANG
+      }
+      updaterResponse(httpResponse(ok, rtn))
+    }
+  }
+}
+
+/**
+ * Get created user tokens.
+ *
+ * @param {string} username - username
+ * @returns {object} - user token
+ */
+const getCreatedTokenOpennebula = (username = '') => {
+  const { now, nowUnix, limitExpirationReuseToken } = setDates()
+  if (username && global?.users?.[username]?.tokens) {
+    let acc = { token: '', time: 0 }
+    global.users[username].tokens.forEach((curr = {}, index = 0) => {
+      const tokenExpirationTime = parseInt(curr.time, 10)
+
+      // this delete expired tokens of global.users[username]
+      if (tokenExpirationTime <= nowUnix) {
+        delete global.users[username].tokens[index]
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We detected that you are using `delete` to remove items from an array. If we use `delete`, it will delete the object property, but will not reindex the array or update its length. It leaves `undefined` holes in the array. Therefore, we recommend you to use `splice` to remove items from an array.

[Learn more](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete)